### ISSUE TITLE
`flex_attention` on `torch.jagged` nested inputs appears to allow cross-sequence attention unless an explicit nested block mask is provided

### ISSUE THREAD

**Issue by ZhiyuanChen** [pytorch/pytorch#177378]
## Summary

For `torch.jagged` nested inputs, `flex_attention(query, key, value)` without an explicit `block_mask` appears to produce different results from `flex_attention(..., block_mask=create_nested_block_mask(...))`, even when the block mask is just "allow everything within each sequence".

That suggests the default nested path may not be isolating per-sequence attention boundaries automatically.

## Why this is surprising

For jagged nested tensors, users generally expect each batch item's ragged sequence to be independent. If no explicit cross-sequence block mask is given, it is surprising if tokens from one sequence can attend to tokens from another sequence inside the stacked jagged representation.

If that behavior is intended, it should be documented clearly. If not, this looks like a correctness bug.

## Repro

```python
import torch
from torch.nn.attention.flex_attention import flex_attention, create_nested_block_mask

device = "cuda"
dtype = torch.float16

def make_nt(lengths, heads=2, dim=8):
    elems = [torch.randn(s, heads, dim, device=device, dtype=dtype) for s in lengths]
    # jagged shape: [B, H, j_seq, D]
    return torch.nested.nested_tensor(elems, layout=torch.jagged).transpose(1, 2)

lengths = [3, 2]
q = make_nt(lengths)
k = make_nt(lengths)
v = make_nt(lengths)

allow_all_same_sequence = create_nested_block_mask(
    lambda b, h, q_idx, kv_idx: torch.ones_like(q_idx, dtype=torch.bool),
    B=len(lengths),
    H=None,
    q_nt=q,
    kv_nt=k,
    _compile=False,
)

out_default = flex_attention(q, k, v)
out_masked = flex_attention(q, k, v, block_mask=allow_all_same_sequence)

print("max diff:", (out_default.values() - out_masked.values()).abs().max().item())
print("allclose:", torch.allclose(out_default.values(), out_masked.values(), atol=5e-2, rtol=5e-2))
```

## Actual result

`out_default` and `out_masked` are not equal.

In one local run:

```text
max diff: 1.234375
allclose: False
```

## Expected result

One of these should be true:

1. `flex_attention(q, k, v)` on jagged nested inputs should already respect sequence boundaries, in which case it should match the explicit "allow everything within each sequence" nested block mask.
2. If the default behavior is intentionally over the fully stacked jagged dimension, that should be documented clearly, since it is not an obvious or intuitive semantic for nested inputs.

## Notes

- This is about eager behavior, not compilation.
- The discrepancy is specifically between:
  - no `block_mask`
  - a nested block mask that only enforces same-sequence boundaries and otherwise allows all attention

### Versions

PyTorch version: 2.10.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.12.12 | packaged by conda-forge | (main, Oct 22 2025, 23:25:55) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-151-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100 80GB HBM3
Nvidia driver version: 565.57.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  224
On-line CPU(s) list:                     0-223
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Platinum 8480+
CPU family:                              6
Model:                                   143
Thread(s) per core:                      2
Core(s) per socket:                      56
Socket(s):                               2
Stepping:                                8
CPU(s) scaling MHz:                      45%
CPU max MHz:                             3800.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4000.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               5.3 MiB (112 instances)
L1i cache:                               3.5 MiB (112 instances)
L2 cache:                                224 MiB (112 instances)
L3 cache:                                210 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-55,112-167
NUMA node1 CPU(s):                       56-111,168-223
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] flake8==7.3.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.3.5
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnx==1.20.1
[pip3] onnx-ir==0.1.15
[pip3] onnxscript==0.6.0
[pip3] pytorch-triton==3.5.1
[pip3] torch==2.10.0+cu126
[pip3] torchao==0.16.0.dev20260205+cu126
[pip3] torchdata==0.11.0
[pip3] torcheval==0.0.7
[pip3] torchmetrics==1.8.2
[pip3] torchtitan==0.2.1
[pip3] torchvision==0.25.0+cu126
[pip3] transformer_engine_torch==2.11.0
[pip3] triton==3.6.0
[conda] numpy                       2.3.5                     pypi_0                pypi
[conda] nvidia-cublas-cu12          12.6.4.1                  pypi_0                pypi
[conda] nvidia-cuda-cupti-cu12      12.6.80                   pypi_0                pypi
[conda] nvidia-cuda-nvrtc-cu12      12.6.77                   pypi_0                pypi
[conda] nvidia-cuda-runtime-cu12    12.6.77                   pypi_0                pypi
[conda] nvidia-cudnn-cu12           9.10.2.21                 pypi_0                pypi
[conda] nvidia-cufft-cu12           11.3.0.4                  pypi_0                pypi
[conda] nvidia-curand-cu12          10.3.7.77                 pypi_0                pypi
[conda] nvidia-cusolver-cu12        11.7.1.2                  pypi_0                pypi
[conda] nvidia-cusparse-cu12        12.5.4.2                  pypi_0                pypi
[conda] nvidia-cusparselt-cu12      0.7.1                     pypi_0                pypi
[conda] nvidia-nccl-cu12            2.27.5                    pypi_0                pypi
[conda] nvidia-nvjitlink-cu12       12.6.85                   pypi_0                pypi
[conda] nvidia-nvtx-cu12            12.6.77                   pypi_0                pypi
[conda] pytorch-triton              3.5.1                     pypi_0                pypi
[conda] torch                       2.10.0+cu126              pypi_0                pypi
[conda] torchao                     0.16.0.dev20260205+cu126  pypi_0                pypi
[conda] torchdata                   0.11.0                    pypi_0                pypi
[conda] torcheval                   0.0.7                     pypi_0                pypi
[conda] torchmetrics                1.8.2                     pypi_0                pypi
[conda] torchtitan                  0.2.1                     pypi_0                pypi
[conda] torchvision                 0.25.0+cu126              pypi_0                pypi
[conda] transformer-engine-torch    2.11.0                    pypi_0                pypi
[conda] triton                      3.6.0                     pypi_0                pypi

cc @cpuhrsch @jbschlosser @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ @chauhang @penguinwu @Chillee @yanboliang @BoyuanFeng @liangel-02 @howardzhang-cv

**Comment by drisspg** [comment]
This is expected we are not planning to support this
